R/module_teal_lockfile.R

```diff
@@ -0,0 +1,219 @@
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application's environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return `NULL`
+#'
+#' @name module_teal_lockfile
+#' @rdname module_teal_lockfile
+#'
+#' @keywords internal
+NULL
+
+#' @rdname module_teal_lockfile
+ui_teal_lockfile <- function(id) {
+  ns <- NS(id)
+  shiny::tagList(
+    tags$span("", id = ns("lockFileStatus")),
+    shinyjs::disabled(downloadLink(ns("lockFileLink"), "Download lockfile"))
+  )
+}
+
+#' @rdname module_teal_lockfile
+srv_teal_lockfile <- function(id) {
+  moduleServer(id, function(input, output, session) {
+    logger::log_debug("Initialize srv_teal_lockfile.")
+    enable_lockfile_download <- function() {
+      shinyjs::html("lockFileStatus", "Application lockfile ready.")
+      shinyjs::hide("lockFileStatus", anim = TRUE)
+      shinyjs::enable("lockFileLink")
+      output$lockFileLink <- shiny::downloadHandler(
+        filename = function() {
+          "renv.lock"
+        },
+        content = function(file) {
+          file.copy(lockfile_path, file)
+          file
+        },
+        contentType = "application/json"
+      )
+    }
+    disable_lockfile_download <- function() {
+      warning("Lockfile creation failed.", call. = FALSE)
+      shinyjs::html("lockFileStatus", "Lockfile creation failed.")
+      shinyjs::disable("lockFileLink")
+    }
+
+    shiny::onStop(function() {
+      if (file.exists(lockfile_path) && !shiny::isRunning()) {
+        logger::log_debug("Removing lockfile after shutting down the app")
+        file.remove(lockfile_path)
+      }
+    })
+
+    lockfile_path <- "teal_app.lock"
+    mode <- getOption("teal.lockfile.mode", default = "")
+    user_lockfile_path <- getOption("teal.lockfile.path", default = "")
+
+    if (!(mode %in% c("auto", "true", "false", "user"))) {
+      stop("'teal.lockfile.mode' option can only be one of \"auto\", \"true\", \"false\", or \"user\". ")
+    }
+
+    if (mode == "false") {
+      logger::log_debug("'teal.lockfile.mode' option is set to 'false'. Hiding lockfile download button.")
+      shinyjs::hide("lockFileLink")
+      return(NULL)
+    }
+
+    if (file.exists(lockfile_path)) {
+      logger::log_debug("Lockfile has already been created for this app - skipping automatic creation.")
+      enable_lockfile_download()
+      return(NULL)
+    }
+
+    if (mode %in% c("auto", "true") && !.is_lockfile_deps_installed()) {
+      warning("Automatic lockfile creation disabled. `mirai` and `renv` packages must be installed.")
+      shinyjs::hide("lockFileLink")
+      return(NULL)
+    }
+
+    if (mode == "auto" && .is_disabled_lockfile_scenario()) {
```

Comment:
this `if` should be placed before previous `if`. Imagine that it is auto-disabled and renv isn't installed - then warning "mirai and renv packages must be installed" but the true reason is that this process would be disabled anyway.

Reply:
totally, thanks!

Reply:
https://github.com/insightsengineering/teal/pull/1263/commits/5bdd79ca3139b9328ddfbc4c0b3faf78bc224841